### DISCUSSION TITLE
Strip the leading dollar sign when copying shell command

### DISCUSSION THREAD

**Discussion by bartekpacia** [community/community#35615] [Other Feature Feedback, Questions, & Ideas]
When writing READMEs (or basically any Markdown) I really like to prefix the commands that I want the reader to copy-paste into their terminal with a dollar sign `$`.

For example, I prefer this:

```
$ gh repo view leancodepl/patrol
```

to this:

```
gh repo view leancodepl/patrol
```

Unfortunately, prefixing the commands with the dollar sign makes it harder for the reader to use the built-in Copy button that GitHub displays by default on the right side of the code snippet.

This often results in small, annoying mistakes:

```
$ $ gh repo view leancodepl/patrol
zsh: command not found: $
```

I'd like the leading dollar sign to be automatically stripped when copying the snippet. Maybe it should happen only when the type of the code snippet is `bash` or `sh` – I don't know.

A site that already does this is Google's docs. [See example here](https://firebase.google.com/docs/cli#initialize_a_firebase_project).

**Comment by Alhadis** [comment]
> Maybe it should happen only when the type of the code snippet is `bash` or `sh` – I don't know.

There's actually a dedicated syntax for shell-sessions:

~~~markdown
```console
$ echo Foo
Foo
```
~~~

The syntax highlighting being used even tokenises the prompt-symbol for the purposes of excluding it from copied code-snippets. However, GitHub have yet to make use of it...

Note that `#` is a legal prompt-symbol here, so comments need to be written a specific way:

~~~markdown
```console
# id -u
0
# # Comment 1
$ # Comment 2
```
~~~

**Comment by bartekpacia** [comment]
Thanks, I didn't know about `console`. 

I'll use it in my Markdowns so that I'll be ready when GitHub implements this feature.

**Reply by zm-cttae** [reply]
I've upvoted but also commenting to +1 this, and explain a use case:

**As a:** consumer/end-user of GitHub projects<br>**I want to:** copy text into my console<br>**So that:** I can copy fixes shared by developers to the console, without knowing much about _terminal/shell_ consoles.

Use case: https://github.com//Gimly/vscode-matlab/pull/174#issuecomment-1435925969

**Comment by mattmc3** [comment]
Users can also solve this problem themselves via their own config. For example, Zsh users can create a dirt simple pass-thru dollar function like so:

```zsh
# .zshrc
function \$ { "$@"; }
```

Bash users can't make a function called `$`, but if they are using [ble.sh](https://github.com/akinomyoga/ble.sh) they can modify the command before it executes:

```bash
# .bashrc
ble/function#advice around ble/widget/default/accept-line '
  if [[ "${_ble_edit_str:0:2}" == "$ " ]]; then
    ble/widget/beginning-of-logical-line
    ble/widget/insert-string "${_ble_edit_str:2}"
    ble/widget/kill-forward-logical-line
  fi
  ble/function#advice/do
'
```

And in Fish, you can add your own paste scrubber:

```fish
function fish_paste_scrubber
    # run the Fish default
    fish_clipboard_paste
    # remove any leading dollar signs from the pasted command
    commandline -r (string replace -r '^\$ ' '' (commandline))
end
```

Then you have to bind it to your paste key sequence in `fish_user_key_bindings`:

```fish
function fish_user_key_bindings
    bind \cv fish_paste_scrubber
    # ... other keybindings ...
end
```

It'd still be a good idea to implement this feature. Just thought it worth mentioning this can be solved on the user end too. Since things like GitLab and BitBucket and blogs and articles etc exist, users will forever be annoyed by this problem if they don't apply a fix themselves.

**Reply by bartekpacia** [reply]
Interesting workarounds, thanks for sharing!

Though I would still like to see GitHub officially supporting this.

**Comment by xvpc** [comment]
Interesting I don't think that can be omitted, thought you can still just modify it in a way to make look prettier.

**Comment by mminot-yseop** [comment]
This discussion could be made a bit more general, as I’m pretty sure AsciiDoc in GitHub has the same issue. Some parts of their code are probably language-agnostic.

By the way, GitHub’s “Copy” button on snippets from pages written in AsciiDoc copy [the callout numbers](https://docs.asciidoctor.org/asciidoc/latest/verbatim/callouts/) (`<1>`, etc.) even though they are not supposed to be part of the code, which often leads to syntax errors when you paste the resulting code. 😰 I got paranoid and now I add code-comment markers before them just in case, now. 😐

**Comment by gene-ressler** [comment]
It's a simple add that makes a difference. I have worked in a proprietary documentation environment that elided dollar sign prompts in Copy button text if the Markdown code block was type bash, shell, etc. We used this a lot for instructions to perform common on-call tasks, where getting things right without hassle was important.

**Comment by bartekpacia** [comment]
Years have passed, and there's still no good solution.

I'll write down all that I have tried. This is such a tiny and stupid problem that GitHub doesn't care about it.

Let's say I want to show in my README.md how to run a command. I use the `shell` syntax highlighting style:

```shell
$ echo "XD"
```

The "copy" button is now broken because it copies the dollar sign. But there's another problem. I also want to show the output of running the command:

```shell
$ echo "XD"
XD
```

The above is okayish, but what if I a long command invocation that is multiline?

```shell
$ echo "XD" \
  --some-flag
XD
```

It's hard to see where the flags end and where the output begins. (Of course, I can find the last `\` and know that the line after it is the last line of command invocation, but it's annoying).

Of course, these are toy examples, but I consider the core problem to be unsolved.
